This patch prioritizes LCP control packets by assigning them to highest priority

IPQoS patch for LCP packet prioritization

diff --git a/drivers/net/ppp/ppp_generic.c b/drivers/net/ppp/ppp_generic.c
--- a/drivers/net/ppp/ppp_generic.c
+++ b/drivers/net/ppp/ppp_generic.c
@@ -1243,6 +1243,20 @@ ppp_send_frame(struct ppp *ppp, struct s
 		/* peek at outbound CCP frames */
 		ppp_ccp_peek(ppp, skb, 0);
 		break;
+#ifdef CONFIG_LANTIQ_IPQOS
+//LCP prioritization hack	
+	case PPP_LCP:
+	case PPP_IPCP:
+	case PPP_PAP:
+	case PPP_CHAP:
+		/* MARK LCP frames with highest priority */
+		/* hack for QOS: if QOS is enabled, give highest priority to LCP control packets */
+		skb->priority = 0;
+#ifdef CONFIG_NETWORK_EXTMARK
+    SET_DATA_FROM_MARK_OPT(skb->extmark, QUEPRIO_MASK, QUEPRIO_START_BIT_POS, 1);
+#endif
+		break;
+#endif
 	}
 
 	/* try to do packet compression */
@@ -1580,6 +1594,10 @@ ppp_channel_push(struct channel *pch)
 	if (pch->chan) {
 		while (!skb_queue_empty(&pch->file.xq)) {
 			skb = skb_dequeue(&pch->file.xq);
+			skb->priority = 0; /* For LCP Prioritization */
+#ifdef CONFIG_NETWORK_EXTMARK
+    	SET_DATA_FROM_MARK_OPT(skb->extmark, QUEPRIO_MASK, QUEPRIO_START_BIT_POS, 1);
+#endif
 			if (!pch->chan->ops->start_xmit(pch->chan, skb)) {
 				/* put the packet back and try again later */
 				skb_queue_head(&pch->file.xq, skb);
@@ -1787,6 +1805,20 @@ ppp_receive_nonmp_frame(struct ppp *ppp,
 	case PPP_CCP:
 		ppp_ccp_peek(ppp, skb, 1);
 		break;
+#ifdef CONFIG_LANTIQ_IPQOS
+// LCP prioritization hack	
+	case PPP_LCP:
+	case PPP_IPCP:
+	case PPP_PAP:
+	case PPP_CHAP:
+		/* MARK LCP frames with highest priority */
+		/* hack for QOS: if QOS is enabled, give highest priority to LCP control packets */
+		skb->priority = 0;
+#ifdef CONFIG_NETWORK_EXTMARK
+    SET_DATA_FROM_MARK_OPT(skb->extmark, QUEPRIO_MASK, QUEPRIO_START_BIT_POS, 1);
+#endif
+		break;
+#endif
 	}
 
 	++ppp->stats64.rx_packets;
